src/hotspot/share/utilities/rbTree.hpp

```diff
@@ -196,52 +196,41 @@ class AbstractRBTree {
   DEBUG_ONLY(mutable bool _expected_visited);
 
 private:
-  template <typename CMP, typename RET, typename ARG1, typename ARG2, typename = void>
-  struct has_cmp_type : std::false_type {};
-  template <typename CMP, typename RET, typename ARG1, typename ARG2>
-  struct has_cmp_type<CMP, RET, ARG1, ARG2, decltype(static_cast<RET(*)(ARG1, ARG2)>(CMP::cmp), void())> : std::true_type {};
-
   template <typename CMP>
-  static constexpr bool HasKeyComparator = has_cmp_type<CMP, RBTreeOrdering, K, K>::value;
+  static constexpr bool HasKeyComparator =
+      std::is_invocable_r_v<RBTreeOrdering, decltype(&CMP::cmp), K, K>;
```

Comment:
Note that these are static data member templates, which were not permitted by the style
guide until about an hour ago. :)